dozer/cogs/filter.py

```diff
@@ -0,0 +1,266 @@
+`{prefix}filter whitelist` - See all of the whitelisted roles
+`{prefix}filter whitelist add Administrators` - Make the Administrators role whitelisted for the filter.
+`{prefix}filter whitelist remove Moderators` - Make the Moderators role no longer whitelisted."""
+
+	@guild_only()
+	@has_permissions(manage_guild=True)
+	@filter.command()
+	async def add(self, ctx, pattern, friendly_name=None):
+		"""Add a pattern to the filter using RegEx. Any word can be added and is tested case-insensitive."""
+		new_filter = WordFilter(guild_id=ctx.guild.id, pattern=pattern, friendly_name=friendly_name or pattern)
+		with db.Session() as session:
+			session.add(new_filter)
+		embed = discord.Embed()
+		embed.title = "Filter {} added".format(new_filter.id)
+		embed.description = "A new filter with the name `{}` was added.".format(friendly_name or pattern)
+		embed.add_field(name="Pattern", value="`{}`".format(pattern))
+		await ctx.send(embed=embed)
+		self.load_filters(ctx.guild.id)
+	add.example_usage = "{prefix}filter whitelist add Swear"
+
+	@guild_only()
+	@has_permissions(manage_guild=True)
+	@filter.command()
+	async def remove(self, ctx, id):
+		"""Remove a pattern from the filter list."""
+		with db.Session() as session:
+			result = session.query(WordFilter).filter(WordFilter.id == id).one_or_none()
+			if result is None:
+				await ctx.send("Filter ID {} not found!".format(id))
+				return
+			if result.guild_id != ctx.guild.id:
+				await ctx.send("That Filter does not belong to this guild.")
+				return
+			result.enabled = False
+			await ctx.send("Filter `{}` with name `{}` deleted.".format(result.id,result.friendly_name))
+			self.load_filters(ctx.guild.id)
+	remove.example_usage = "{prefix}filter remove 7"
+
+	@guild_only()
+	@has_permissions(manage_guild=True)
+	@filter.command(name="dm")
+	async def dm_config(self, ctx, config: bool):
+		"""Set whether filter words should be DMed when used in bot messages"""
+		with db.Session() as session:
+			result = session.query(WordFilterSetting).filter(WordFilterSetting.guild_id == ctx.guild.id,
+															 WordFilterSetting.setting_type == "dm")\
+				.one_or_none()
+			try:
+				before_setting = result.value
+				result.value = config
+			except AttributeError:
+				before_setting = None
+				result = WordFilterSetting(guild_id=ctx.guild.id, setting_type="dm", value=config)
+				session.add(result)
+			await ctx.send("The DM setting for this guild has been changed from {} to {}.".format(before_setting, result.value))
+	dm_config.example_usage = "{prefix}filter dm_config True"
+
+	@guild_only()
+	@filter.group(invoke_without_command=True)
+	async def whitelist(self, ctx):
+		"""List all whitelisted roles for this server"""
+		with db.Session() as session:
+			results = session.query(WordFilterRoleWhitelist).filter(WordFilterRoleWhitelist.guild_id == ctx.guild.id).all()
+			roles_text = ""
+			for db_role in results:
```

Comment:
This looks like another good place to use `str.join` rather than repeated concatenation.